package NYT.NControllerAgent.NProto;

import "yt/ytlib/controller_agent/proto/job.proto";

import "yt_proto/yt/client/node_tracker_client/proto/node.proto";

import "yt_proto/yt/core/misc/proto/guid.proto";
import "yt_proto/yt/core/misc/proto/error.proto";

////////////////////////////////////////////////////////////////////////////////

message TAllocationInfo
{
    required NYT.NProto.TGuid allocation_id = 1;
}

message TClusterNetworkBandwidthAvailability
{
    message TAvailability
    {
        optional string cluster_name = 1;
        optional bool is_available = 2;
    }

    repeated TAvailability availability = 1;
}

// Sends job statistics to controller agent.
message TReqHeartbeat
{
    required uint32 node_id = 1;
    required NNodeTrackerClient.NProto.TNodeDescriptor node_descriptor = 2;
    required NYT.NProto.TGuid controller_agent_incarnation_id = 3;

    repeated TJobStatus jobs = 4;
    repeated NYT.NProto.TGuid unconfirmed_job_ids = 9;
    required int32 confirmed_job_count = 10;

    repeated TAllocationInfo allocations = 11;

    optional uint64 sequence_number = 12;

    optional TClusterNetworkBandwidthAvailability cluster_network_bandwidth_availability = 13;
}

message TJobToAbort
{
    required NYT.NProto.TGuid job_id = 1;
    optional int32 abort_reason = 2;
    optional bool graceful = 3 [default = false];
    optional bool request_new_job = 4 [default = false];
}

message TJobToInterrupt
{
    required NYT.NProto.TGuid job_id = 1;
    required int64 timeout = 2;
    required int32 reason = 3; // NScheduler::EInterruptionReason
}

message TJobToStore
{
    required NYT.NProto.TGuid job_id = 1;
}

message TJobToConfirm
{
    required NYT.NProto.TGuid job_id = 1;
}

message TRspHeartbeat
{
    repeated TJobToStore jobs_to_store = 1;
    repeated TJobToRemove jobs_to_remove = 2;
    repeated TJobToConfirm jobs_to_confirm = 3;
    repeated TJobToAbort jobs_to_abort = 4;
    repeated TJobToInterrupt jobs_to_interrupt = 6;

    repeated NYT.NProto.TGuid unknown_operation_ids = 5;

    reserved 7;
}

////////////////////////////////////////////////////////////////////////////////

message TReqSettleJob
{
    required uint32 node_id = 1;
    required NNodeTrackerClient.NProto.TNodeDescriptor node_descriptor = 2;

    required NYT.NProto.TGuid controller_agent_incarnation_id = 3;

    required NYT.NProto.TGuid allocation_id = 4;
    optional NYT.NProto.TGuid last_job_id = 6;
    required NYT.NProto.TGuid operation_id = 5;
}

message TRspSettleJob
{
    message TBriefJobInfo
    {
        required NYT.NProto.TGuid job_id  = 1;
    }

    oneof job_info_or_error {
        NYT.NProto.TError error = 1;
        TBriefJobInfo job_info = 2;
    }
    // The actual spec is passed via attachments.
}

////////////////////////////////////////////////////////////////////////////////
